src/storage/mvcc/txn.rs

```diff
@@ -619,6 +621,204 @@ impl<S: Snapshot> MvccTxn<S> {
+            key_prefix: vec![],
+            remove_older: false,
+
+            level: 0,
+            versions: 0,
+            rows: 0,
+            stale_versions: 0,
+            deleted: 0,
+            skipped: 0,
+            start: Instant::now_coarse(),
+        }
+    }
+
+    fn delete_default_key(&self, key: &[u8]) {
+        self.write_batch.delete(key).unwrap();
+        if self.write_batch.data_size() > DEFAULT_DELETE_BATCH_SIZE {
+            let mut opts = WriteOptions::new();
+            opts.set_sync(false);
+            self.db.write_opt(&self.write_batch, &opts).unwrap();
+            self.write_batch.clear();
+        }
+    }
+}
+
+impl Drop for WriteCompactionFilter {
+    fn drop(&mut self) {
+        if !self.write_batch.is_empty() {
+            let mut opts = WriteOptions::new();
+            opts.set_sync(true);
+            self.db.write_opt(&self.write_batch, &opts).unwrap();
+        }
+        info!(
+            "WriteCompactionFilter uses {}s", self.start.elapsed_secs();
+            "output_level" => self.output_level,
+            "level" => self.level,
+            "versions" => self.versions,
+            "stale_versions" => self.stale_versions,
+            "rows" => self.rows,
+            "deleted" => self.deleted,
+            "skipped" => self.skipped,
+        );
+        GC_DELETED_VERSIONS.inc_by(self.deleted as i64);
+    }
+}
+
+impl CompactionFilter for WriteCompactionFilter {
+    fn filter(
+        &mut self,
+        level: usize,
+        key: &[u8],
+        value: &[u8],
+        _: &mut Vec<u8>,
+        _: &mut bool,
+    ) -> bool {
+        self.level = level;
+        let safe_point = self.safe_point.load(Ordering::Acquire);
+        if safe_point == 0 {
+            return false;
+        }
+
+        let (key_prefix, commit_ts) = match Key::split_on_ts_for(key) {
+            Ok((key, ts)) => (key, ts),
+            // Invalid MVCC keys, don't touch them.
+            Err(_) => return false,
+        };
+
+        self.versions += 1;
+        if self.key_prefix != key_prefix {
+            self.key_prefix.clear();
+            self.key_prefix.extend_from_slice(key_prefix);
+            self.remove_older = false;
+        }
+
+        if commit_ts <= safe_point {
+            self.stale_versions += 1;
+        }
+
+        let mut filtered = self.remove_older;
+        let (write_type, start_ts, has_short) = Write::parse_without_value(value).unwrap();
+        if !self.remove_older {
+            // here `filtered` must be false.
+            match write_type {
+                WriteType::Rollback | WriteType::Lock => filtered = true,
+                WriteType::Delete => {
+                    self.remove_older = true;
+                    filtered =
+                        self.output_level == self.max_level && self.key_prefix != self.max_delete;
+                }
+                WriteType::Put => {
+                    self.remove_older = true;
```

Comment:
The last version before safepoint should always be kept (except `Delete`). You may need to return false early before here, as long as `commit_ts > safe_point`.